src/exec/use_pty/pipe.rs

```diff
@@ -100,16 +100,100 @@ impl<L: Read + Write + AsRawFd, R: Read + Write + AsRawFd> Pipe<L, R> {
     }
 }
 
-/// The size of the internal buffer of the pipe.
-const BUFSIZE: usize = 6 * 1024;
+struct RingBuffer {
+    storage: [u8; Self::LEN],
+    // The start index of the non-empty section of the buffer.
+    start: usize,
```

Comment:
in this implementation, the `start` and `len` field need to withhold the invariant that they stay in the range `[0, LEN]` (I think `start` needs be *less* than `LEN` but `len` can be equal to `LEN`?) so you could add some debug assertions at the stand and the end of `remove` and `insert` to try to catch bugs that break those invariants.